File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleButton.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/**
 * The full-width primary action. Pill radius, and a minimum height rather than a fixed one so the
 * label wraps instead of clipping at large font scales.
 *
 * Loading is a distinct state from disabled: the button stops accepting taps but keeps its enabled
 * colours, because a form that greys out mid-submission reads as having rejected the input.
 */
@Composable
fun UseSmileIDSampleButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    loading: Boolean = false,
    testId: String? = null,
) {
    Button(
        onClick = onClick,
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = SmileDimens.sizeControlLg)
            .tagged(testId),
        enabled = enabled && !loading,
        shape = RoundedCornerShape(SmileDimens.radiusControl),
        colors = ButtonDefaults.buttonColors(
            containerColor = UseSmileIDSampleTheme.colors.button.primaryBackground,
            contentColor = UseSmileIDSampleTheme.colors.button.primaryText,
            disabledContainerColor = if (loading) {
                UseSmileIDSampleTheme.colors.button.primaryBackground
            } else {
                UseSmileIDSampleTheme.colors.button.disabledBackground
            },
            disabledContentColor = if (loading) {
                UseSmileIDSampleTheme.colors.button.primaryText
            } else {
                UseSmileIDSampleTheme.colors.button.disabledText
            },
        ),
        contentPadding = ButtonDefaults.ContentPadding,
    ) {
        Box(contentAlignment = Alignment.Center) {
            if (loading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(SmileDimens.sizeIconMd),
                    color = UseSmileIDSampleTheme.colors.button.primaryText,
                    strokeWidth = SmileDimens.borderWidthThick,
                )
            } else {
                Text(
                    text = text,
                    style = UseSmileIDSampleTheme.type.buttonFont,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(vertical = SmileDimens.space4),
                )
            }
        }
    }
}

File: flutter/sample_ui/lib/src/components/use_smileid_sample_button.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';

/// The full-width primary action; a minimum height so the label wraps rather than clips, and
/// loading refuses taps while keeping the enabled colours.
class UseSmileIDSampleButton extends StatelessWidget {
  /// [loading] outranks [enabled]: a button mid-submission is not a disabled button.
  const UseSmileIDSampleButton({
    required this.text,
    required this.onPressed,
    this.enabled = true,
    this.loading = false,
    this.testId,
    super.key,
  });

  /// The label.
  final String text;

  /// What a tap does; never called while loading or disabled.
  final VoidCallback onPressed;

  /// Whether the action is available.
  final bool enabled;

  /// Whether a submission is in flight.
  final bool loading;

  /// The `sample_*` id the screen supplies, since most ids are assigned per screen.
  final String? testId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(context);
    final bool tappable = enabled && !loading;
    final Color background = tappable || loading
        ? colors.button.primaryBackground
        : colors.button.disabledBackground;
    final Color foreground = tappable || loading
        ? colors.button.primaryText
        : colors.button.disabledText;

    final Widget button = Semantics(
      button: true,
      enabled: tappable,
      identifier: testId,
      child: Material(
        color: background,
        borderRadius: UseSmileIDSampleShapes.pill,
        child: InkWell(
          onTap: tappable ? onPressed : null,
          borderRadius: UseSmileIDSampleShapes.pill,
          child: ConstrainedBox(
            constraints: const BoxConstraints(minHeight: SmileDimens.sizeControlLg),
            child: Padding(
              padding: const EdgeInsets.symmetric(
                horizontal: SmileDimens.spacingLg,
                vertical: SmileDimens.space4,
              ),
              child: Center(
                child: loading
                    ? SizedBox(
                        width: SmileDimens.sizeIconMd,
                        height: SmileDimens.sizeIconMd,
                        child: CircularProgressIndicator(
                          strokeWidth: SmileDimens.borderWidthThick,
                          valueColor: AlwaysStoppedAnimation<Color>(foreground),
                        ),
                      )
                    : Text(
                        text,
                        textAlign: TextAlign.center,
                        style: UseSmileIDSampleType.buttonFont.copyWith(color: foreground),
                      ),
              ),
            ),
          ),
        ),
      ),
    );
    return SizedBox(width: double.infinity, child: button);
  }
}
